cppcache/test/util/functionalTests.cpp

```diff
@@ -32,4 +32,12 @@ TEST(string, geode_hash) {
   EXPECT_EQ(48, hash("0"));
   EXPECT_EQ(57, hash("9"));
   EXPECT_EQ(1077910243, hash("supercalifragilisticexpialidocious"));
+
+  EXPECT_EQ(1544552287, hash("You had me at meat tornad\u00F6!\U000F0000"));
+
+  auto str = std::string("You had me at");
+  str.push_back(0);
+  str.append("meat tornad\u00F6!\U000F0000");
```

Comment:
What are you doing with the string here?  Some indicator (comment?) about why you're putting a double null in the middle would help

Reply:
Where is the double null? I see one NUL character.

Reply:
declaration of the string null-terminates, then push_back(0) gives you a second one, right?

Reply:
Nope! `std::string` is not null terminated. The `const char *` passed into the construct is null terminated it effectively results in `char[strlen(cstr)] chars; memcpy(chars, cstr, strlen(cstr));` (obviously not C++ code, just an illustration).

Reply:
What is happening here is that we are working around the issue that C++ won't let you initialize a `std::string` that does have an embedded NUL character in the sequence, since it effectively uses `strlen` functions to initialize the internal storage with the given c-string. The result is that we have one NUL and some other unicode characters that cause Java Modified UTF-8 to produce bytes that are not really UTF-8 compatible (by design).